File: app/src/main/java/com/six/cat/sixcat/base/BaseActivity.java
package com.six.cat.sixcat.base;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

import com.six.cat.sixcat.R;
import com.six.cat.sixcat.module.base.IBasePresenter;
import com.six.cat.sixcat.module.base.IBaseView;
import com.six.cat.sixcat.utils.ActivityManager;
import com.trello.rxlifecycle2.LifecycleTransformer;
import com.trello.rxlifecycle2.android.ActivityEvent;
import com.trello.rxlifecycle2.android.FragmentEvent;
import com.trello.rxlifecycle2.components.support.RxAppCompatActivity;

import butterknife.ButterKnife;
import butterknife.Unbinder;

public abstract class BaseActivity<T extends IBasePresenter> extends RxAppCompatActivity implements IBaseView<T> {

    public ActivityManager mManager;
    private Unbinder bind;
    protected T presenter;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getLayoutId());
        bind = ButterKnife.bind(this);
        setPresenter(presenter);
        initView(savedInstanceState);
        initToolBar();
        mManager = ActivityManager.getInstance();
        mManager.addActivity(this);
    }

    public abstract int getLayoutId();

    protected abstract void initView(Bundle savedInstanceState);

    protected abstract void initToolBar();

    @Override
    public <T> LifecycleTransformer<T> bindToLife() {
        return bindUntilEvent(ActivityEvent.DESTROY);
    }

}

File: app/src/main/java/com/six/cat/sixcat/views/base/BaseActivity.kt
package com.six.cat.sixcat.views.base

import android.os.Bundle
import butterknife.ButterKnife
import butterknife.Unbinder
import com.six.cat.sixcat.presenter.IBasePresenter
import com.six.cat.sixcat.presenter.IBaseView
import com.six.cat.sixcat.utils.ActivityManager
import com.trello.rxlifecycle2.LifecycleTransformer
import com.trello.rxlifecycle2.android.ActivityEvent
import com.trello.rxlifecycle2.components.support.RxAppCompatActivity

abstract class BaseActivity<T : IBasePresenter> : RxAppCompatActivity(), IBaseView<T> {

    var mManager: ActivityManager? = null
    private var bind: Unbinder? = null
    protected var mSetPresenter: T? = null

    abstract fun getLayoutId(): Int


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(getLayoutId())
        bind = ButterKnife.bind(this)
        mSetPresenter(mSetPresenter)
        initView(savedInstanceState)
        initToolBar()
        mManager = ActivityManager.getInstance()
        mManager!!.addActivity(this)
    }

    protected abstract fun initView(savedInstanceState: Bundle?)

    protected abstract fun initToolBar()

    override fun <T> bindToLife(): LifecycleTransformer<T> {
        return bindUntilEvent(ActivityEvent.DESTROY)
    }

}
